src/main/java/io/r2dbc/h2/codecs/ValueLobHandler.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.h2.codecs;
+
+import io.r2dbc.spi.Blob;
+import io.r2dbc.spi.Clob;
+import org.h2.value.Value;
+import org.h2.value.ValueLob;
+import org.h2.value.ValueLobDb;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.SynchronousSink;
+import reactor.core.scheduler.Schedulers;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Generic handler for H2 {@link ValueLobDb}/{@link ValueLob}s.
+ * <p>
+ * NOTE: Both modern and legacy BLOB/CLOB H2 types extend {@link Value#getInputStream()}} virtually, allowing this class to handle both.
+ */
+abstract class ValueLobHandler<T> {
+
+    private final Value lobDb;
+
+    private SynchronousSink<T> valueLobHandlerSink;
+
+    /**
+     * Capture the {@link ValueLobDb}/{@link ValueLob} as a generic {@link Value}, after verifying it's a Lob.
+     */
+    ValueLobHandler(Value value) {
+        if (!(value instanceof ValueLobDb) && !(value instanceof ValueLob)) {
+            throw new IllegalStateException("value must be ValueLobDb or ValueLob!");
+        }
+        this.lobDb = value;
+    }
+
+    /**
+     * Open the {@link ValueLobDb}/{@link ValueLob}'s {@link InputStream} and pipe the bytes into a {@link Flux}.
+     */
+    Flux<T> valueLobToFlux() {
+        return Flux.<T, InputStream>generate(
+            this.lobDb::getInputStream,
+            (source, sink) -> {
+                this.valueLobHandlerSink = sink;
+                try {
+                    // Buffer that is divisible by 2, 3, and 5, ensuring any character combo will fit.
```

Comment:
I don't understand this comment.

This code is used not only with BLOB values, but also with CLOBs?

CLOBs in H2 are stored internally in UTF-8 encoding. It's one of variable-length encodings. I don't see where and how do you handle incomplete characters at the end of the chuck. Am I missing something?

Reply:
I found that Chinese characters require 3 bytes to encode while English characters only need 2. It would break if the total stream of bytes was longer than the length of the array, and if that caused the last character to get split up. By picking a buffer length divisible by 2, 3, and 5, the break never happens in the middle of a character's encoding.

I can verify if this holds using UTF-8 based on earlier comments.

Reply:
In UTF-8 characters are encoded with 1, 2, 3, or 4 bytes. CLOBs may and will contain characters with different encoded length, for example, many national characters use 2 or more bytes, but digits, space, colon, semicolon, point, comma, etc. use only 1 byte. A break will always be possible with any length of array. You need to handle it properly.

As an alternative you can try to work with `char` buffers instead. But characters outside of BMP also use two `char`s in Java with high and low surrogate that produce a complete character together. I don't know how CLOBs are used in R2DBC, but if such chunks are passed to applications it's better not to break chunks between pair of surrogate characters. If R2DBC API does not expose these buffers and provides only a complete streams, it does not matter, because such characters will be glued together again.